src/libraries/System.Private.Uri/src/System/Uri.cs

```diff
@@ -1129,42 +1142,51 @@ public string DnsSafeHost
 
                 EnsureHostString(false);
 
-                if (_info.DnsSafeHost != null)
-                {
-                    // Cached
-                    return _info.DnsSafeHost;
-                }
-                else if (_info.Host!.Length == 0)
+                if (_info.Host!.Length == 0)
                 {
                     // Empty host, no possible processing
                     return string.Empty;
                 }
 
+                // If DnsSafeHost is different from the host, DnsSafeHost will be the same as IdnHost
```

Comment:
What guarantees that?  I'm not quite following the logic.

Reply:
And if that is true, doesn't that mean we should swap the implementation such that the logic to compute this is in IdnHost, and then DnsSafeHost just returns either Host or IdnHost?

Reply:
DnsSafeHost = Host, unless it's IPV6 or Basic.
IdnHost = DnsSafeHost, unless it's DNS

As the conditions of them being modified don't overlap:
`DnsSafeHost != Host` => `IdnHost == DnsSafeHost`
And if `DnsSafeHost == Host`, it's cached anyway since `Host` is cached

> And if that is true, doesn't that mean we should swap the implementation such that the logic to compute this is in IdnHost

Yes, we can do that. Might be easier to follow 👍